dependency-specification.rst

```diff
@@ -0,0 +1,324 @@
+behaviour as in for string in Python if the operator exists in Python. For
+those operators which are not defined in Python, the result should be False.
+
+User supplied constants are always encoded as strings with either ``'`` or
+``"`` quote marks. Note that backslash escapes are not defined, but existing
+implementations do support them them. They are not included in this
+specification because none of the variables to which constants can be compared
+contain quote-requiring values.
+
+The variables in the marker grammar such as "os_name" resolve to values looked
+up in the Python runtime. With the exception of "extra" all values are defined
+on all Python versions today - it is an error in the implementation of markers
+if a value is not defined.
+
+Unknown variables must raise a SyntaxError rather than resulting in a
+comparison that evaluates to True or False.
+
+The "extra" variable is special. It is used by wheels to signal which
+specifications apply to a given extra in the wheel ``METADATA`` file, but
+since the ``METADATA`` file is based on a draft version of PEP-426, there is
+no current specification for this. Regardless, outside of a context where this
+special handling is taking place, the "extra" variable should result in a
+SyntaxError like all other unknown variables.
+
+.. list-table::
+   :header-rows: 1
+
+   * - Marker
+     - Python equivalent
+     - Sample values
+   * - ``os_name``
+     - ``os.name``
+     - ``posix``, ``java``
+   * - ``sys_platform``
+     - ``sys.platform``
+     - ``linux``, ``darwin``, ``java1.8.0_51``
+   * - ``platform_machine``
+     - ``platform.machine()``
+     - ``x86_64``
+   * - ``python_implementation``
+     - ``platform.python_implementation()``
+     - ``CPython``, ``Jython``
+   * - ``platform_version``
+     - ``platform.version()``
+     - ``#1 SMP Fri Apr 25 13:07:35 EDT 2014``
+
+       ``Java HotSpot(TM) 64-Bit Server VM, 25.51-b03, Oracle Corporation``
+
+       ``Darwin Kernel Version 14.5.0: Wed Jul 29 02:18:53 PDT 2015; root:xnu-2782.40.9~2/RELEASE_X86_64``
+   * - ``python_version``
+     - ``platform.python_version()[:3]``
+     - ``3.4``, ``2.7``
+   * - ``python_full_version``
+     - see definition below
+     - ``3.4.0``, ``3.5.0b1``
+   * - ``extra``
+     - A SyntaxError except when defined by the context interpreting the
+       specification.
+     - ``test``
+
+Backwards Compatibility
+=======================
+
+Most of this PEP is already widely deployed and thus offers no compatibiltiy
+concerns.
+
+There are however a few points where the PEP differs from the deployed base.
+
+Firstly, PEP-440 direct references haven't actually been deployed in the wild,
+but they were designed to be compatibly added, and there are no known
+obstacles to adding them to pip or other tools that consume the existing
+dependency metadata in distributions - particularly since they won't be
+permitted to be present in PyPI uploaded distributions anyway.
```

Comment:
I agree with @dstufft that the "support in this spec" and the "support in pip" questions can be separated for direct dependencies. Depending on software that has yet to be formally published is an interesting problem, as it has clear practical benefits *within* an organisation and during development, but poses challenges for dependencies *between* organisations and in released versions of software.

PEP 426 tries to formulate that distinction in terms of "public index server" vs "private index server", but I'm not entirely sure that's the right way of thinking about it. Fortunately, I think we can safely punt on it in relation to this particular spec - we know we want an answer to "How are direct dependencies written?", even though we're not entirely sure on "When we do want to accept direct dependencies on unpublished versions during installation?".